api/v1alpha1/sparkcluster_webhook.go

```diff
@@ -51,57 +64,93 @@ func (r *SparkCluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
+		sc.Spec.TCPWorkerWebPort = sparkDefaultWorkerWebPort
 	}
-	if r.Spec.DashboardPort == 0 {
+	if sc.Spec.TCPMasterWebPort == 0 {
+		log.Info("setting default master web port", "value", sparkDefaultMasterWebPort)
+		sc.Spec.TCPMasterWebPort = sparkDefaultMasterWebPort
+	}
+	if sc.Spec.DashboardPort == 0 {
 		log.Info("setting default dashboard port", "value", sparkDefaultDashboardPort)
-		r.Spec.DashboardPort = sparkDefaultDashboardPort
+		sc.Spec.DashboardPort = sparkDefaultDashboardPort
+	}
+	if sc.Spec.DashboardServicePort == 0 {
+		log.Info("setting default dashboard service port", "value", sparkDefaultDashboardServicePort)
+		sc.Spec.DashboardServicePort = sparkDefaultDashboardServicePort
 	}
-	if r.Spec.EnableDashboard == nil {
+	if sc.Spec.EnableDashboard == nil {
 		log.Info("setting enable dashboard flag", "value", *sparkDefaultEnableDashboard)
-		r.Spec.EnableDashboard = sparkDefaultEnableDashboard
+		sc.Spec.EnableDashboard = sparkDefaultEnableDashboard
 	}
-	if r.Spec.NetworkPolicy.Enabled == nil {
+	if sc.Spec.NetworkPolicy.Enabled == nil {
 		log.Info("setting enable network policy flag", "value", *sparkDefaultEnableNetworkPolicy)
-		r.Spec.NetworkPolicy.Enabled = sparkDefaultEnableNetworkPolicy
+		sc.Spec.NetworkPolicy.Enabled = sparkDefaultEnableNetworkPolicy
 	}
-	if r.Spec.NetworkPolicy.ClientServerLabels == nil {
+	if sc.Spec.NetworkPolicy.ExternalPolicyEnabled == nil {
+		log.Info("setting enable external network policy flag", "value", *sparkDefaultEnableExternalNetworkPolicy)
+		sc.Spec.NetworkPolicy.ExternalPolicyEnabled = sparkDefaultEnableExternalNetworkPolicy
+	}
+	if sc.Spec.NetworkPolicy.ClientServerLabels == nil {
 		log.Info("setting default network policy client labels", "value", sparkDefaultNetworkPolicyClientLabels)
-		r.Spec.NetworkPolicy.ClientServerLabels = sparkDefaultNetworkPolicyClientLabels
+		sc.Spec.NetworkPolicy.ClientServerLabels = sparkDefaultNetworkPolicyClientLabels
 	}
-	if r.Spec.NetworkPolicy.DashboardLabels == nil {
+	if sc.Spec.NetworkPolicy.DashboardLabels == nil {
 		log.Info("setting default network policy dashboard labels", "value", sparkDefaultNetworkPolicyClientLabels)
-		r.Spec.NetworkPolicy.DashboardLabels = sparkDefaultNetworkPolicyClientLabels
+		sc.Spec.NetworkPolicy.DashboardLabels = sparkDefaultNetworkPolicyClientLabels
 	}
-	if r.Spec.Worker.Replicas == nil {
+	if sc.Spec.Worker.Replicas == nil {
 		log.Info("setting default worker replicas", "value", *sparkDefaultWorkerReplicas)
-		r.Spec.Worker.Replicas = sparkDefaultWorkerReplicas
+		sc.Spec.Worker.Replicas = sparkDefaultWorkerReplicas
 	}
-
-	if r.Spec.Image == nil {
-		log.Info("setting default image", "value", *sparkDefaultImage)
-		r.Spec.Image = sparkDefaultImage
+	if sc.Spec.Worker.WorkerMemoryLimit == "" {
+		log.Info("setting default worker memory request", "value", sparkDefaultWorkerMemoryLimit)
+		sc.Spec.Worker.WorkerMemoryLimit = sparkDefaultWorkerMemoryLimit
```

Comment:
I don't know how I feel about setting a default worker memory limit of 4.5Gi. I'd almost say it's better to enforce that users provide a memory limit during validation. Maybe I don't know enough about Spark, and that's the minimum it needs for any workload, but that feels like high, default resource price to pay.